### PR TITLE
[SPARK-37220][SQL] Do not split input file for Parquet reader with aggregate push down

### PR DESCRIPTION
**Pull request by c21**
### What changes were proposed in this pull request?

As a followup of https://github.com/apache/spark/pull/34298/files#r734795801, Similar to ORC aggregate push down, we can disallow split input files for Parquet reader as well. See original comment for more details of motivation. Also fix the string of `RowDataSourceScanExec` to only print out `PushedAggregates` and `PushedGroupby`, to be aligned with `PushedLimit` and `PushedSample`, as there's not so many queries can benefit from aggregate push down, so we don't need to print those unnecessary information.

### Why are the changes needed?

Avoid unnecessary file splits in multiple tasks for Parquet reader with aggregate push down.

### Does this PR introduce _any_ user-facing change?

No.

### How was this patch tested?

Existing unit test in `FileSourceAggregatePushDownSuite.scala`.

### CODE REVIEW

**Comment by c21** [comment]
@huaxingao, @sunchao and @viirya - could you help take a look when you have time? Thanks.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 1]
```diff
@@ -151,9 +151,10 @@ case class RowDataSourceScanExec(
 
     Map(
       "ReadSchema" -> requiredSchema.catalogString,
-      "PushedFilters" -> seqToString(markedFilters.toSeq),
-      "PushedAggregates" -> aggString,
-      "PushedGroupby" -> groupByString) ++
+      "PushedFilters" -> seqToString(markedFilters.toSeq)) ++
+      pushedDownOperators.aggregation.fold(Map[String, String]()) { v =>
```
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:155]: hmm is this change necessary? are `aggString` and `groupByString` still used afterwards?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetScan.scala [hunk 2]
```diff
@@ -47,7 +47,11 @@ case class ParquetScan(
     pushedAggregate: Option[Aggregation] = None,
     partitionFilters: Seq[Expression] = Seq.empty,
     dataFilters: Seq[Expression] = Seq.empty) extends FileScan {
-  override def isSplitable(path: Path): Boolean = true
+  override def isSplitable(path: Path): Boolean = {
+    // If aggregate is pushed down, only the file footer will be read once,
+    // so file should not be split across multiple tasks.
+    pushedAggregate.isEmpty
```
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetScan.scala:53]: 👍

**Review by sunchao — COMMENTED** [review]
Thanks @c21, looks good just one nit.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 1 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:148]: @sunchao - forget to remove `aggString` and `groupByString`. Removed. I think it'd better to align with `PushedLimit` and `PushedSample`. It's a minor change so packed here. I can make another minor PR if needed.

Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:148]: Got it. This looks fine to me.

Comment (huaxingao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:148]: The change looks good. I plan to fix this but didn't have a chance to do it yet. One problem, though. In the tests I check if the aggregate pushes down by checking the keywords in explain. If the explain contains `PushedAggregation: []`, then the aggregate is not pushed down. Since now we don't have the `[]`, some of the tests will fail. We probably should change all the tests to use a different method (collect aggregates in optimizedPlan) to check if aggregate push down. Probably do this in a separate PR.

**Review by viirya — COMMENTED** [review]
Looks okay.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 1 above)
Comment (sunchao) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:148]: IMO it may be better to keep `[]` for empty aggregate/groupby, to be consistent with `PushedFilters` and those in `OrcScan`/`ParquetScan`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala [hunk 3]
```diff
@@ -151,9 +144,10 @@ case class RowDataSourceScanExec(
...
+        Map("PushedAggregates" -> seqToString(v.aggregateExpressions),
+          "PushedGroupby" -> seqToString(v.groupByColumns))} ++
```
Comment (sunchao) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala:150]: nit: `PushedGroupBy`.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala (hunk 1 above)
Comment (c21) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/DataSourceScanExec.scala]: @huaxingao - ah yes I forgot about the test. I am leaving `RowDataSourceScanExec` as it is for now, and we can discuss later, thanks for comment @sunchao and @huaxingao.

**Review by viirya — APPROVED** [review]

**Review by sunchao — APPROVED** [review]

**Comment by c21** [comment]
Thank you @sunchao, @huaxingao and @viirya for review!

**Comment by HyukjinKwon** [comment]
Quick question on:

> Existing unit test in `FileSourceAggregatePushDownSuite.scala.`

How did the existing tests pass before this PR?

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetScan.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetScan.scala:53]: Oh, okay. Got it now.

Comment (HyukjinKwon) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetScan.scala:53]: Oh, okay. Got it now.

**Review by HyukjinKwon — COMMENTED** [review]
LGTM2

File: sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetPartitionReaderFactory.scala [hunk 4]
```diff
@@ -92,11 +92,6 @@ case class ParquetPartitionReaderFactory(
       ParquetFooterReader.readFooter(conf, filePath, SKIP_ROW_GROUPS)
     } else {
       // For aggregate push down, we will get max/min/count from footer statistics.
-      // We want to read the footer for the whole file instead of reading multiple
-      // footers for every split of the file. Basically if the start (the beginning of)
-      // the offset in PartitionedFile is 0, we will read the footer. Otherwise, it means
-      // that we have already read footer for that file, so we will skip reading again.
-      if (file.start != 0) return null
```
Comment (c21) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/parquet/ParquetPartitionReaderFactory.scala:99]: > Quick question on: Existing unit test in FileSourceAggregatePushDownSuite.scala. How did the existing tests pass before this PR?

@HyukjinKwon - I think we are in the same page based on your latest comment, but just to be noisy here in case anything is missing. Before this PR, when a single file is split into multiple splits across multiple tasks, we have the logic here to only process the split of file if `file.start == 0`, so only the first split of file will be processed, and every file is processed only once. So here is the trick.
